Add documentation - Getting Started : Deployment (#1832)
kuzzleio/kuzzle: doc/2/guides/getting-started/8-deploy/index.md

title: Deploy
description: Deploy your application on a remote server
order: 800
---

# Deploy

Kuzzle is just a Node.js application that also needs Elasticsearch and Redis to run.  

The only specifity is that Kuzzle needs to compile C and C++ dependencies so the `npm install` will also be looking for `python`, `make` and `g++` packages.   

At Kuzzle we like to use Docker and Docker Compose to quickly deploy applications.  
In this guide we will see how to deploy a Kuzzle application on a remote server.

**Prerequisites:**
  - A SSH access to a remote server running on Linux

>   - [Docker](https://docs.docker.com/engine/install/) and [Docker Compose](https://docs.docker.com/compose/install/) installed on this server

Reviewer: I think we really need to add "A Kuzzle custom application" as a prerequisite to this guide (not sure about the wording)

Author: Since it's still the getting started people are supposed to have a small application by now

Reviewer: Yes but if they find this page using the Algolia search it can be confusing to don't know this prerequisite

::: info
In this guide we will perform a basic deployment of a Kuzzle application.  

> For production deployments, we strongly recommend to deploy your application with the [cluster](/core/2/guides/some-link) mode to take advantage of high availability and scalability.  

Reviewer 2: ```suggestion
For production deployments, we strongly recommend to deploy your application in [cluster mode](/core/2/guides/some-link) to take advantage of high availability and scalability.  
```

    command: node /var/app/app.js
    restart: always
    container_name: kuzzle
    ports:
      - "7512:7512"
      - "1883:1883"
    depends_on:
      - redis
      - elasticsearch
    environment:
      - kuzzle_services__storageEngine__client__node=http://elasticsearch:9200
      - kuzzle_services__internalCache__node__host=redis
      - kuzzle_services__memoryStorage__node__host=redis
      - NODE_ENV=production

  redis:
    image: redis:5
    command: redis-server --appendonly yes
    restart: always
    volumes:
      - redis-data:/data

  elasticsearch:
    image: kuzzleio/elasticsearch:7
    restart: always
    ulimits:
      nofile: 65536
    volumes:
      - es-data:/usr/share/elasticsearch/data

volumes:
  es-data:
    driver: local
  redis-data:
    driver: local
```

This configuration allows to run a Kuzzle application with Node.js alongside Elasticsearch and Redis.

Kuzzle needs compiled dependencies for the cluster and the realtime engine.  

We are going to use a multi-stage Docker file to build the dependencies and then use the [node:12-stretch-slim](https://hub.docker.com/_/node?tab=description) image to run the application.

> Create the `deployment/kuzzle.dockerfile` file with the following content:

Author: ```suggestion
Create the `docker/kuzzle.dockerfile` file with the following content:
```

```dockerfile
# builder image
FROM node:12-stretch-slim as builder

RUN  set -x \
  && apt-get update && apt-get install -y \
       curl \
       make \
       g++ \
       python \
       libzmq3-dev

ADD . /var/app

WORKDIR /var/app

RUN  npm install --production

# run image
FROM node:12-stretch-slim

COPY --from=builder /var/app /var/app
```

## Deploy on your remote server

Now we are going to use `scp` (SSH copy) to copy our application on the remote server.  

```bash
$ scp -r ../playground <user>@<server-ip>:.
```

Then simply connect to your server and run your application with Docker Compose:

```bash
$ ssh <user>@<server-ip>

[...]

$ docker-compose -f docker/docker-compose.yml up -d
```

Your Kuzzle application is now up and running on port 7512!

Learn more about:

>  - [Deployment](/core/2/api/some-links)

Reviewer: I think we could add the SSL setup guide link (again)
